process/elasticproc/tokens/sovereignIndexTokensHandler.go

```diff
@@ -0,0 +1,166 @@
+	"strings"
+	"time"
+
+	"github.com/elastic/go-elasticsearch/v7"
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/data/esdt"
+
+	"github.com/multiversx/mx-chain-es-indexer-go/client"
+	"github.com/multiversx/mx-chain-es-indexer-go/client/logging"
+	"github.com/multiversx/mx-chain-es-indexer-go/data"
+	indexerdata "github.com/multiversx/mx-chain-es-indexer-go/process/dataindexer"
+	"github.com/multiversx/mx-chain-es-indexer-go/process/elasticproc"
+	"github.com/multiversx/mx-chain-es-indexer-go/process/elasticproc/converters"
+	"github.com/multiversx/mx-chain-es-indexer-go/process/elasticproc/factory"
+)
+
+type sovereignIndexTokensHandler struct {
+	indexingEnabled        bool
+	mainChainElasticClient elasticproc.DatabaseClientHandler
+	esdtPrefix             string
+}
+
+// NewSovereignIndexTokensHandler creates a new sovereign index tokens handler
+func NewSovereignIndexTokensHandler(mainChainElastic factory.ElasticConfig, esdtPrefix string) (*sovereignIndexTokensHandler, error) {
+	var mainChainElasticClient elasticproc.DatabaseClientHandler
+	if mainChainElastic.Enabled {
+		var err error
+		argsEsClient := elasticsearch.Config{
+			Addresses:     []string{mainChainElastic.Url},
+			Username:      mainChainElastic.UserName,
+			Password:      mainChainElastic.Password,
+			Logger:        &logging.CustomLogger{},
+			RetryOnStatus: []int{http.StatusConflict},
+			RetryBackoff:  retryBackOff,
+		}
+		mainChainElasticClient, err = client.NewElasticClient(argsEsClient)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &sovereignIndexTokensHandler{
+		indexingEnabled:        mainChainElastic.Enabled,
+		mainChainElasticClient: mainChainElasticClient,
+		esdtPrefix:             esdtPrefix,
+	}, nil
+}
+
+func retryBackOff(attempt int) time.Duration {
+	return time.Duration(math.Exp2(float64(attempt))) * time.Second
+}
+
+// IndexCrossChainTokens will index the new tokens properties
+func (sit *sovereignIndexTokensHandler) IndexCrossChainTokens(elasticClient elasticproc.DatabaseClientHandler, scrs []*data.ScResult, buffSlice *data.BufferSlice) error {
+	if !sit.indexingEnabled {
+		return nil
+	}
+
+	notFoundTokens, err := sit.getTokensFromScrs(elasticClient, scrs)
+	if err != nil {
+		return err
+	}
+
+	if len(notFoundTokens) == 0 { // no new tokens
+		return nil
+	}
+
+	// get tokens from main chain elastic db
+	mainChainTokens := &data.ResponseTokenInfo{}
+	err = sit.mainChainElasticClient.DoMultiGet(context.Background(), notFoundTokens, indexerdata.TokensIndex, true, mainChainTokens)
+	if err != nil {
+		return err
+	}
+
+	return sit.indexNewTokens(mainChainTokens.Docs, buffSlice)
+}
+
+func (sit *sovereignIndexTokensHandler) getTokensFromScrs(elasticClient elasticproc.DatabaseClientHandler, scrs []*data.ScResult) ([]string, error) {
+	receivedTokensIDs := make([]string, 0)
+	for _, scr := range scrs {
+		if scr.SenderShard == core.MainChainShardId {
+			for _, token := range scr.Tokens {
```

Comment:
Please no imbricated `for`. Create a separate func for the second one

Reply:
done